### PR TITLE
SCC | Add SCC (`openshift.io/required-scc`) Annotations (core and endpoint)

### PR DESCRIPTION
**Pull request by shirady**
### Describe the Problem
Epic: [RHSTOR-8757](https://redhat.atlassian.net/browse/RHSTOR-8757) Enforce least-privileged Security Context Constraint (SCC) across ODF pods.

### Explain the changes
Worked according to the design doc:
1. YAML Manifests (added to the `spec.template.metadata.annotations`, so pods will inherit this):
- `deploy/internal/statefulset-core.yaml` (noobaa-core StatefulSet): Add `openshift.io/required-scc: noobaa-core`
- `deploy/internal/deployment-endpoint.yaml` (noobaa-endpoint Deployment): Add `openshift.io/required-scc: noobaa-endpoint`
2. Go Code Reconcilers:
- `pkg/system/phase2_creating.go`: under `SetDesiredCoreApp ` set the annotation.
- `pkg/system/phase4_configuring.go`: under `SetDesiredDeploymentEndpoint` set the annotation.
3. Add `noobaa-core` to `CollectSCC` (it already had `noobaa-endpoint`).

### Issues: 
1. none

##### Local Cluster:
1. Build the images and install NooBaa system on Rancher Desktop (see [guide](https://github.com/noobaa/noobaa-operator/blob/master/doc/dev_guide/deply_noobaa_on_minikube_or_rancher_desktop.md)).
2. Check that you can see the annotation using the following examples:
- ` kubectl -n test1  get statefulset noobaa-core -o jsonpath='{.spec.template.metadata.annotations}'`
- `kubectl -n test1  get deployment noobaa-endpoint -o jsonpath='{.spec.template.metadata.annotations}'`

#### OCP Cluster
Detailed instructions are in the [comment](https://github.com/noobaa/noobaa-operator/pull/1965#issuecomment-4779109588) below.

- [ ] Doc added/updated
- [ ] Tests added

## Summary by CodeRabbit

* **Chores**
  * Updated OpenShift/Kubernetes manifests and operator-managed templates to explicitly request the appropriate Security Context Constraints (SCC) for NooBaa components, including the endpoint, metrics adapter, agent, core, database, and operator.
  * Hardened SCC configuration by adding `seccompProfiles: runtime/default`.
  * Adjusted core and database SCC behavior (e.g., `runAsUser` and `fsGroup`) and applied `restricted-v2` where applicable to maintain intended security posture.

### CODE REVIEW

File: pkg/bundle/deploy.go [hunk 1]
```diff
@@ -4880,13 +4883,15 @@ spec:
...
 
 const File_deploy_internal_pod_agent_yaml = `apiVersion: v1
 kind: Pod
 metadata:
   labels:
     app: noobaa
+  annotations:
+    openshift.io/required-scc: noobaa-core
   name: noobaa-agent
 spec:
```
[the remark this answers is not in the document]
Comment (shirady) [reply pkg/bundle/deploy.go:4899]: I was able to install the system, so why do you think I can get issues during pod creation?

**Comment by shirady** [comment]
@ezio-auditore could you please take a look?

**Review by dannyzaken — APPROVED** [review]
LGTM. 
These changes need to be tested in openshift, to make sure none of the pods are now failing on SCC

**Comment by shirady** [comment]
## Additional Testing Instructions:
On the local setup, I validated only the existence of the annotation.
Thanks to the comment above, I also tried to validate on the OCP cluster.

1. Used `cluster-bot` with: `launch 4.22 aws` (although ODF is not in version 4.22 yet)
2. Installed MCG standalone (due to limited resources on the cluster-bot)
3. Update the CRD: `kubectl apply -f deploy/crds/noobaa.io_noobaas.yaml`
Note: Otherwise, I have issues with Connect after updating the image due to missing vector status in noobaa status.
4. Update the CSV of mcg with the image of core and operator with the code change.
core image: `noobaa/noobaa-core:master-20260622`
operator image: `quay.io/rh_ee_sdymnik/noobaa-operator:add-annotations-scc-20260623`

**endpoint pod:**
`kubectl get pod -n openshift-storage noobaa-endpoint-8474d4df9f-jxb9c -o yaml | grep scc`

```shell
    openshift.io/required-scc: noobaa-endpoint
    openshift.io/scc: noobaa-endpoint
    security.openshift.io/validated-scc-subject-type: serviceaccount
```

**core-pod:**
`kubectl get pod -n openshift-storage noobaa-core-0 -o yaml | grep scc`

```shell
    openshift.io/required-scc: noobaa-core
    openshift.io/scc: noobaa-core
    security.openshift.io/validated-scc-subject-type: serviceaccount
```

File: deploy/scc.yaml [hunk 2]
```diff
@@ -11,3 +11,5 @@ seLinuxContext:
 supplementalGroups:
   type: RunAsAny
 readOnlyRootFilesystem: true
+seccompProfiles:
```
Comment (dannyzaken) [inline deploy/scc.yaml]: why is this needed? When do we create this SCC?

File: pkg/system/phase2_creating.go [hunk 3]
```diff
@@ -373,6 +373,10 @@ func (r *Reconciler) SetDesiredNooBaaDB() error {
 	if dbAnnotations, ok := r.NooBaa.Spec.Annotations["db"]; ok {
 		NooBaaDB.Spec.Template.Annotations = dbAnnotations
 	}
+	if NooBaaDB.Spec.Template.Annotations == nil {
+		NooBaaDB.Spec.Template.Annotations = make(map[string]string)
+	}
+	NooBaaDB.Spec.Template.Annotations[util.OpenShiftRequiredSCCAnnotation] = "noobaa-db"
```
Comment (dannyzaken) [inline pkg/system/phase2_creating.go]: Regarding the DB - the standalone is not too relevant anymore. We probably need to remove this code sometime soon. You should be able to set annotations on the pods using the `InheritedMetadata` property of the cluster, as we do for the `app: noobaa` label. see example [here](https://github.com/dannyzaken/noobaa-operator/blob/f30a050a5ac24d38e7b1e31c49d7c77e5ce30957/pkg/system/db_reconciler.go#L246-L253)

DB pods should be restricted-v2

File: deploy/scc.yaml (hunk 2 above)
Comment (shirady) [reply deploy/scc.yaml]: Without this, I had an error:
```
  Warning  FailedCreate  67s (x53 over 36m)  replicaset-controller  Error creating: pods "noobaa-operator-6d7cd9fbdc-" is forbidden: unable to validate against any security context constraint: [metadata.annotations[seccomp.security.alpha.kubernetes.io/pod]: Forbidden: seccomp may not be set, metadata.annotations[container.seccomp.security.alpha.kubernetes.io/noobaa-operator]: Forbidden: seccomp may not be set]
```
All the details are inside the testing instructions on the cluster ([link](https://github.com/noobaa/noobaa-operator/pull/1965#issuecomment-4779109588)).

File: pkg/system/phase2_creating.go (hunk 3 above)
Comment (shirady) [reply pkg/system/phase2_creating.go]: I'm not sure I understood your comment.
You think that the DB pods should not have `noobaa-db` on them and should have `restricted-v2`?
I worked according to the design doc.

Comment (ezio-auditore) [reply pkg/system/phase2_creating.go]: Yes should be restriced v2 for new cnpg deployment, please update the change as danny mentions, I will update the design doc

**Review by ezio-auditore — CHANGES_REQUESTED** [review]
Everything else looks good, will there be a second PR for unit tests?

**Comment by shirady** [comment]
> Everything else looks good, will there be a second PR for unit tests?

@ezio-auditore I didn't plan to add unit tests for adding the annotations in a following PR.
What kind of unit tests do you have in mind?

File: pkg/system/phase2_creating.go (hunk 3 above)
Comment (shirady) [reply pkg/system/phase2_creating.go]: I will add the change as it is in the current code for labels, as attached ([link](https://github.com/dannyzaken/noobaa-operator/blob/f30a050a5ac24d38e7b1e31c49d7c77e5ce30957/pkg/system/db_reconciler.go#L246-L253)) for annotations.

**Comment by ezio-auditore** [comment]
> > Everything else looks good, will there be a second PR for unit tests?
> 
> @ezio-auditore I didn't plan to add unit tests for adding the annotations in a following PR. What kind of unit tests do you have in mind?

Unit tests to check all pods deployed under noobaa to have `openshift.io/required-scc` annotation.Can be implemented in followup PR

**Comment by shirady** [comment]
@coderabbitai review

File: pkg/system/phase2_creating.go (hunk 3 above)
Comment (shirady) [reply pkg/system/phase2_creating.go]: I tried to test it on the ODF cluster, and this failed - I tried to delete one of the DB pods (`noobaa-db-pg-cluster-2`), and it was not created by CNPG due to an error. So, it doesn't seem like the correct way to set those annotations.


`kubectl logs -n openshift-storage cnpg-controller-manager-64d9648687-2dpvw | grep -iE 'noobaa-db-pg-cluster|error|failed'`
```shell
{"level":"info","ts":"2026-06-29T11:14:01.523014954Z","msg":"Updating cluster annotations on pod","controller":"cluster","controllerGroup":"postgresql.cnpg.noobaa.io","controllerKind":"Cluster","Cluster":{"name":"noobaa-db-pg-cluster","namespace":"openshift-storage"},"namespace":"openshift-storage","name":"noobaa-db-pg-cluster","reconcileID":"b5c4c41d-4c86-421d-b6d1-0cb2d1c42845","pod":"noobaa-db-pg-cluster-1"}
{"level":"error","ts":"2026-06-29T11:14:01.536030223Z","msg":"while patching instance metadata","controller":"cluster","controllerGroup":"postgresql.cnpg.noobaa.io","controllerKind":"Cluster","Cluster":{"name":"noobaa-db-pg-cluster","namespace":"openshift-storage"},"namespace":"openshift-storage","name":"noobaa-db-pg-cluster","reconcileID":"b5c4c41d-4c86-421d-b6d1-0cb2d1c42845","instanceName":"noobaa-db-pg-cluster-1","error":"pods \"noobaa-db-pg-cluster-1\" is forbidden: invalid change of required security context constraint annotation: openshift.io/required-scc"scc"
```

You can see here that the DB pod was not re-created after deletion:
`kubectl get pods -n openshift-storage | grep -iE 'noobaa|cnpg'`
```
cnpg-controller-manager-64d9648687-2dpvw                          1/1     Running     0          33m
noobaa-core-0                                                     2/2     Running     0          94m
noobaa-db-pg-cluster-2                                            1/1     Running     0          4h
noobaa-endpoint-5d98cdfdbc-tqkzv                                  1/1     Running     0          94m
noobaa-operator-6b96c5df85-wt4m9                                  1/1     Running     0          35m
shira-bs-pv-noobaa-pod-0edaa164                                   1/1     Running     0          170m
```

Update: only when I edited the CSV and edited it to the previous image (without adding the annotation on the DB pods) I could see the 2 DB pods.

Comment (ezio-auditore) [reply pkg/system/phase2_creating.go]: Also instead of setting template's annotations, it should be the annotations of PodTemplateSpec ObjectMeta

Comment (ezio-auditore) [reply pkg/system/phase2_creating.go]: https://github.com/red-hat-storage/ocs-operator/pull/3964/changes#diff-32b7cad897cdf567ab54824c8b0a5fd65057de0ce167c328f6e4f040455973e3L162

Comment (shirady) [reply pkg/system/phase2_creating.go]: I'm not sure I understand you -
For operator, core, endpoint, standalone DB `required-scc` is already set on `spec.template.metadata.annotations`, for example:

`kubectl get deployment -n openshift-storage noobaa-endpoint -o yaml`
```
spec:
...
  template:
    metadata:
      annotations:
...
        openshift.io/required-scc: noobaa-endpoint
```

For the CNPG, we tried to use `InheritedMetadata` which, as stated before, failed.

If you have a specific suggestion, please add it.

Comment (ezio-auditore) [reply pkg/system/phase2_creating.go]: I understand the issue, openshift is not letting any change in scc annotation for cnpg pods. And clearly `InheritedMetadata` has an issue. Lets keep it open as a known issue for now.

File: pkg/system/db_reconciler.go [hunk 4]
```diff
@@ -250,6 +251,10 @@ func (r *Reconciler) reconcileClusterSpec(dbSpec *nbv1.NooBaaDBSpec) error {
 		r.CNPGCluster.Spec.InheritedMetadata.Labels = map[string]string{}
 	}
 	r.CNPGCluster.Spec.InheritedMetadata.Labels["app"] = "noobaa"
+	if r.CNPGCluster.Spec.InheritedMetadata.Annotations == nil {
+		r.CNPGCluster.Spec.InheritedMetadata.Annotations = map[string]string{}
+	}
+	r.CNPGCluster.Spec.InheritedMetadata.Annotations[secv1.RequiredSCCAnnotation] = "restricted-v2"
```
Comment (ezio-auditore) [inline pkg/system/db_reconciler.go]: remove this for now and please open a bug by which we can track

**Review by ezio-auditore — CHANGES_REQUESTED** [review]

File: pkg/system/db_reconciler.go (hunk 4 above)
Comment (shirady) [reply pkg/system/db_reconciler.go]: I will remove it, thanks.

File: deploy/internal/hpav2-deployment-adapter.yaml [hunk 5]
```diff
@@ -23,6 +23,8 @@ spec:
         app.kubernetes.io/component: metrics-adapter
         app.kubernetes.io/name: prometheus-adapter
         app.kubernetes.io/version: 0.10.0
+      annotations:
```
Comment (jackyalbo) [inline deploy/internal/hpav2-deployment-adapter.yaml]: @shirady is this something we really want? Part of the epic? To harden the choosing to restricted-v2 specific (maybe cluster have only restricted for example) - I would double-check - I think that we should require SCC that we are responsible for creating, WDYT?

Comment (shirady) [reply deploy/internal/hpav2-deployment-adapter.yaml]: It is part of the epic, but since you're raising concerns, I will remove it from this PR and double-check it with @ezio-auditore .
In the epic level, this is what they want: to have pre-defined the SCC for all components.
From what I read, since 4.11 all OCP clusters should have the  `restricted-v2` SCC, in this epic there are a couple of components that use `restricted-v2` SCC.

File: deploy/operator.yaml [hunk 6]
```diff
@@ -14,6 +14,8 @@ spec:
       labels:
         app: noobaa
         noobaa-operator: deployment
+      annotations:
```
Comment (jackyalbo) [inline deploy/operator.yaml]: I think also here we don't want to force using restricted-v2

File: deploy/internal/hpav2-deployment-adapter.yaml (hunk 5 above)
Comment (ezio-auditore) [reply deploy/internal/hpav2-deployment-adapter.yaml]: @jackyalbo are you suggesting creating a new custom scc, if that is the case it is acceptable as well, if not we should have `restricted-v2` for now and test it, if bugs arise that would mean we need a different less restrictive scc

File: deploy/operator.yaml (hunk 6 above)
Comment (shirady) [reply deploy/operator.yaml]: I want this PR merged with the agreed SCCs.
Since you're raising a concern about the operator SCC, I will exclude it from this PR.

File: pkg/system/phase2_creating.go (hunk 3 above)
Comment (shirady) [reply pkg/system/phase2_creating.go]: @coderabbitai, could you open a GitHub issue with the gap related to CNPG with the annotation of `openshift.io/required-scc` and include the information from this discussion?

**Review by jackyalbo — APPROVED** [review]

**Review by ezio-auditore — APPROVED** [review]
LGTM
